File: packages/sdk-python/src/stagehand/openai_compatible.py
"""Client-side adapter for OpenAI-compatible Chat Completions endpoints."""

from __future__ import annotations

import asyncio
import json
import urllib.parse
from collections.abc import Mapping
from typing import Any
from urllib.request import Request, urlopen

from ._generated.models import (
    LLMImageContent,
    LLMMessage,
    LLMStructuredGenerateParams,
    LLMStructuredGenerateResult,
    LLMTextContent,
)
from .client_types import LLMGenerateCallback, LLMGenerateInput


def _content(message: LLMMessage) -> str | list[dict[str, Any]]:
    blocks = message.content if isinstance(message.content, list) else [message.content]
    parts: list[dict[str, Any]] = []
    for block in blocks:
        content = block.root
        if isinstance(content, LLMTextContent):
            parts.append({"type": "text", "text": content.text})
        elif isinstance(content, LLMImageContent):
            parts.append({
                "type": "image_url",
                "image_url": {"url": f"data:{content.mime_type};base64,{content.data}"},
            })
        else:
            raise TypeError(f"OpenAI-compatible models do not accept {content.type} content")
    if all(part["type"] == "text" for part in parts):
        return "".join(part["text"] for part in parts)
    return parts


def _post_json(
    url: str, api_key: str, headers: Mapping[str, str], payload: dict[str, Any]
) -> dict[str, Any]:
    request = Request(
        url,
        data=json.dumps(payload).encode(),
        headers={
            **headers,
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        method="POST",
    )
    with urlopen(request, timeout=60) as response:
        return json.load(response)


def open_ai_compatible(
    *,
    model: str,
    base_url: str,
    api_key: str,
    headers: Mapping[str, str] | None = None,
    query_params: Mapping[str, str] | None = None,
    extra_body: Mapping[str, Any] | None = None,
) -> LLMGenerateCallback:
    """Return a callback for an OpenAI-compatible Chat Completions endpoint."""
    url = f"{base_url.rstrip('/')}/chat/completions"
    if query_params:
        query_string = urllib.parse.urlencode(dict(query_params))
        url = f"{url}{'&' if '?' in url else '?'}{query_string}"
    extra_headers = dict(headers or {})
    extra_payload = dict(extra_body or {})

    async def generate(params: LLMGenerateInput) -> LLMStructuredGenerateResult:
        if not isinstance(params, LLMStructuredGenerateParams):
            raise TypeError("Stagehand only issues structured generations")
        response_format = params.response_format
        messages: list[dict[str, Any]] = []
        if params.system_prompt:
            messages.append({"role": "system", "content": params.system_prompt})
        messages.extend(
            {"role": message.role.value, "content": _content(message)}
            for message in params.messages
        )
        json_schema: dict[str, Any] = {
            "name": response_format.name,
            "schema": response_format.schema_.model_dump(by_alias=True)
            if response_format.schema_ is not None
            else None,
            "strict": True,
        }
        if response_format.description is not None:
            json_schema["description"] = response_format.description
        payload: dict[str, Any] = {
            "model": model,
            "messages": messages,
            "response_format": {"type": "json_schema", "json_schema": json_schema},
            **extra_payload,
        }
        if params.temperature is not None and "temperature" not in extra_payload:
            payload["temperature"] = params.temperature
        response = await asyncio.to_thread(_post_json, url, api_key, extra_headers, payload)
        choices = response.get("choices", [])
        text = choices[0].get("message", {}).get("content") if choices else None
        if not isinstance(text, str):
            raise ValueError("OpenAI-compatible response did not include message content")
        return LLMStructuredGenerateResult.model_validate({
            "role": "assistant",
            "content": {"type": "text", "text": text},
            "output_format": "json_schema",
            "structured_content": json.loads(text),
        })

    return generate

File: packages/sdk-ts/src/openAICompatible.ts
import type {
  LLMGenerateParams,
  LLMGenerateResult,
  LLMJsonSchemaResponseFormat,
  LLMMessage,
  LLMMessageContentBlock,
} from "@browserbasehq/stagehand-protocol/types";
import { z } from "zod";
import type { ClientLLM } from "./clientSchemas.js";

export type OpenAICompatibleOptions = {
  model: string;
  baseURL: string;
  apiKey: string;
  headers?: Record<string, string>;
  extraBody?: Record<string, unknown>;
};

type ChatContentPart =
  | { type: "text"; text: string }
  | { type: "image_url"; image_url: { url: string } };

type ChatMessage = {
  role: string;
  content: string | ChatContentPart[];
};

type ChatCompletionResponse = {
  choices?: Array<{ message?: { content?: string | null } }>;
};

function contentBlocks(content: LLMMessage["content"]): LLMMessageContentBlock[] {
  return Array.isArray(content) ? content : [content];
}

function chatContent(message: LLMMessage): ChatMessage["content"] {
  const parts: ChatContentPart[] = [];
  for (const block of contentBlocks(message.content)) {
    if (block.type === "text") {
      parts.push({ type: "text", text: block.text });
      continue;
    }
    if (block.type === "image") {
      parts.push({
        type: "image_url",
        image_url: { url: `data:${block.mimeType};base64,${block.data}` },
      });
      continue;
    }
    throw new TypeError(`OpenAI-compatible models do not accept ${block.type} content`);
  }

  if (parts.every((part) => part.type === "text")) {
    return parts.map((part) => part.text).join("");
  }
  return parts;
}

function chatMessages(params: LLMGenerateParams): ChatMessage[] {
  const messages: ChatMessage[] = [];
  if (params.systemPrompt) {
    messages.push({ role: "system", content: params.systemPrompt });
  }
  for (const message of params.messages) {
    messages.push({ role: message.role, content: chatContent(message) });
  }
  return messages;
}

function structuredFormat(responseFormat: LLMJsonSchemaResponseFormat) {
  return {
    type: "json_schema" as const,
    json_schema: {
      name: responseFormat.name,
      description: responseFormat.description,
      schema: responseFormat.schema,
      strict: true,
    },
  };
}

function completionText(body: ChatCompletionResponse): string {
  const text = body.choices?.[0]?.message?.content;
  if (typeof text !== "string") {
    throw new Error("OpenAI-compatible response did not include message content");
  }
  return text;
}

/** A client-side model for any endpoint that implements OpenAI Chat Completions. */
export function openAICompatible(options: OpenAICompatibleOptions): ClientLLM {
  const endpoint = `${options.baseURL.replace(/\/+$/, "")}/chat/completions`;
  const headers = new Headers(options.headers);
  headers.set("authorization", `Bearer ${options.apiKey}`);
  headers.set("content-type", "application/json");

  return {
    async generate(params: LLMGenerateParams): Promise<LLMGenerateResult> {
      if (params.responseFormat?.type !== "json_schema") {
        throw new TypeError("Stagehand only issues structured generations");
      }

      const body: Record<string, unknown> = {
        ...options.extraBody,
        model: options.model,
        messages: chatMessages(params),
        response_format: structuredFormat(params.responseFormat),
      };
      if (params.temperature !== undefined && !("temperature" in body)) {
        body.temperature = params.temperature;
      }

      const response = await fetch(endpoint, {
        method: "POST",
        headers,
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        throw new Error(
          `OpenAI-compatible request failed (${response.status}): ${await response.text()}`,
        );
      }

      const text = completionText((await response.json()) as ChatCompletionResponse);
      return {
        role: "assistant",
        content: { type: "text", text },
        outputFormat: "json_schema",
        structuredContent: z.json().parse(JSON.parse(text)),
      };
    },
  };
}
